hyperspy/tests/utils/test_roi.py

```diff
@@ -486,6 +486,37 @@ def test_line2droi_angle(self):
         with pytest.raises(ValueError):
             r.angle(axis='z')
 
+    def test_repr_None(self):
+        # Setting the args=None sets them as traits.Undefined, which didn't 
+        # have a string representation in the old %s style.
+        repr(Point1DROI())
+        repr(Point2DROI())
+        repr(Line2DROI())
+        repr(RectangularROI())
+        repr(SpanROI())
+        repr(CircleROI())
+
+    def test_repr_vals(self):
+        repr(Point1DROI(1.1))
+        repr(Point2DROI(1.1, 2.1))
+        repr(Line2DROI(0, 0, 1, 1, 0.1))
+        repr(RectangularROI(0, 0, 1, 1))
+        repr(SpanROI(3., 5.))
+        repr(CircleROI(5, 5, 3))
+        repr(CircleROI(5, 5, 3, 1))
+
+    def test_undefined_call(self):
+        rect = RectangularROI(None, None, None, None)
+        with pytest.raises(AttributeError, match='not yet been set'):
+            rect(self.s_s)
+
+    def test_undefined_call_circle(self):
+        circ1 = CircleROI(None, None, None, None)
+        with pytest.raises(AttributeError, match='not yet been set'):
+            circ1(self.s_s)
+        circ2 = CircleROI(5, 5, 1, None)
+        circ2(self.s_s) # r_inner as undefined should not raise error
+
```

Comment:
Testing the `add_widget` method would be good too.
It should be possible to put this in a for loop as in comment above.

Reply:
How about like this?
```python
    def test_widget_initialisation(self):
        self.s_s.plot()
        for roi in [Point1DROI, Point2DROI, RectangularROI, SpanROI, Line2DROI, CircleROI]:
            r = roi()
            r._set_default_values(self.s_s)
            r.add_widget(self.s_s)
```